### ISSUE TITLE
Inner product queried format is not optimum

### ISSUE THREAD

**Issue by CaoZhongZ** [intel/mkl-dnn#464]
Suppose we have a weights with IO formats, if we do the query with format::any trick and we could get a suggested OI format. But the time of reordering from IO to OI and computing the inner product doesn't make up just calculate it in IO. Should we still keep the query routine in inner product?

**Comment by nathan-greeneltch-intel** [comment]
I'm sorry @CaoZhongZ . I don't understand your question. Can you restate?

**Comment by uyongw** [comment]
It seems in most of 2d cases, IO is faster than OI even don't counting the memory reordering. This is perhaps because io is friendly for vectorization. I saw there is a gather/tranpose for oi weights. So I doubt whether the query returns the best format.

Secondly, there is no ic-first format for 3d and 5d weights. That says, to match ncw, we need iwo, to match ncdhw we need idhwo but they are defined from format tag.

https://github.com/intel/mkl-dnn/blob/master/src/common/inner_product_pd.hpp#L148
```
        if (src_md.format_kind == format_kind::any) {
            CHECK(memory_desc_init_by_tag(src_md,
                        utils::pick(ndims() - 2, nc, ncw, nchw, ncdhw)));
        }
        if (dst_md.format_kind == format_kind::any)
            CHECK(memory_desc_init_by_tag(dst_md, nc));
        if (weights_md.format_kind == format_kind::any) {
            CHECK(memory_desc_init_by_tag(weights_md,
                        utils::pick(ndims() - 2, oi, oiw, oihw, oidhw)));
        }
```

**Comment by CaoZhongZ** [comment]
@emfomenk From what @uyongw observed, the query results is wrong and there isn't enough benefit to fix on certain formats. Should we remove the query parts of the inner product?

**Comment by shelleygoel** [comment]
@CaoZhongZ  _"But the time of reordering from IO to OI and computing the inner product doesn't make up just calculate it in IO. "_
It would be nice if you could present some data to support this conclusion. Is this something that you observe in topology training time? In addition MKLDNN_VERBOSE output would be helpful too.

InnerProduct is compute intensive, so I imagine reorder time would be a small fraction of the total time of "reorder + innerProduct".  There are several important questions to ask here:
1) What is the computation time for  `IO` vs `OI` ?
2) How many times does this reorder happen during topology training/inference? For example for inference, we can reorder weights once and cache them. Also for training, we can reorder initial weights from `IO` to `OI` and then  train the weights in the format `OI` and then reorder them once at the end if desired. 
3) It also helps to specify the hardware, data type etc.

**Comment by shelleygoel** [comment]
@uyongw `nc`, `oihw` are  the default formats for Inner product, which are used as a fallback if for some reason the optimal implementations cannot be used. They might not be the formats returned by the inner product primitive. The optimal format mainly depend on the hardware, tensor sizes and data types. So it's difficult to say what optimal formats would be without knowing hardware, tensor sizes and data types specified.

**Comment by uyongw** [comment]
@nathan-greeneltch-intel This is to discuss the performance issue of inner-product primitive we met in framework integration. We found the suggested data format queried from mkldnn is not optimal.

@shelleygoel I agree. However it seems to me current implementation does not check hardware and tensor shape. BTW, do we have plan to switch to mkldnn JIT implementation (instead of mkl blas) or to support blocked format (say Oi16o for example)?

**Comment by densamoilov** [comment]
@uyongw, you mentioned that in most 2d cases the IO format is better for the inner product, but what the cases are? Could you provide us with those?  Actually, the IO (and other formats derived from it) is used in the inner product for int8 data type while the inner product for fp32 relies on the OI, so we may want to consider switching to IO format for fp32 in the future if we can benefit from it.

We will remove the dependency on MKL BLAS and rely on our own implementations of MKL BLAS functions, but this will not add support of blocked formats. Do you need the blocked formats to avoid reordering prior inner product or there is something else?

**Comment by uyongw** [comment]
@densamoilov 

Summary: 
1. MKL BLAS based IP: OI weights is faster on large batch size, IO weights is faster on single batch size.
2. JIT GEMM based IP: OI weights is normally faster than IO weights.
3. JIT GEMM based IP is faster for larger batch-size, slower on single batch size (if this is a performance bug, please fix it).

Hardware
```
$ lscpu
Thread(s) per core:    2
Core(s) per socket:    28
Socket(s):             2
Model name:            Intel(R) Xeon(R) Platinum 8180 CPU @ 2.50GHz

```
Test cases
```
$ cat inputs/ip/t.x
mb42ic1024oc3072
mb42ic1024oc1024
mb42ic1024oc4096
mb42ic4096oc1024
mb1ic1024oc3072
mb1ic1024oc1024
mb1ic1024oc4096
mb1ic4096oc1024
```
Bench
Extend benchdnn a little bit. Please check with the patch below:
[extend benchdnn's inner-product for testability](https://github.com/uyongw/mkl-dnn/commit/5588edb4f12b7ddd4b25ca857acde64066fbd699)




Result
```
# mkl blas, weights: oi
$ OMP_NUM_THREADS=28 KMP_HW_SUBSET=1s,28c,1t KMP_AFFINITY=compact,granularity=fine ./benchdnn --ip --mode=P --batch=inputs/ip/t.x
perf,42,3072,1024,1,1,1,"wip",FWD_B,f32,0.493896,535.013,0.502499,525.854
perf,42,1024,1024,1,1,1,"wip",FWD_B,f32,0.32251,273.109,0.327671,268.807
perf,42,4096,1024,1,1,1,"wip",FWD_B,f32,0.65625,536.871,0.666971,528.241
perf,42,1024,4096,1,1,1,"wip",FWD_B,f32,1.3125,268.435,1.32193,266.521
perf,1,3072,1024,1,1,1,"wip",FWD_B,f32,0.0163574,384.624,0.0173675,362.256
perf,1,1024,1024,1,1,1,"wip",FWD_B,f32,0.00756836,277.095,0.00903827,232.03
perf,1,4096,1024,1,1,1,"wip",FWD_B,f32,0.0209961,399.532,0.0221443,378.815
perf,1,1024,4096,1,1,1,"wip",FWD_B,f32,0.0200195,419.021,0.0225358,372.236

# mkl blas, weights: io
$ OMP_NUM_THREADS=28 KMP_HW_SUBSET=1s,28c,1t KMP_AFFINITY=compact,granularity=fine ./benchdnn --ip --mode=P --batch=inputs/ip/t.x
perf,42,3072,1024,1,1,1,"wip",FWD_B,f32,0.289062,914.132,0.292472,903.476
perf,42,1024,1024,1,1,1,"wip",FWD_B,f32,0.187988,468.542,0.190455,462.474
perf,42,4096,1024,1,1,1,"wip",FWD_B,f32,0.390869,901.38,0.395638,890.515
perf,42,1024,4096,1,1,1,"wip",FWD_B,f32,0.74292,474.239,0.750047,469.733
perf,1,3072,1024,1,1,1,"wip",FWD_B,f32,0.0275879,228.051,0.0318088,197.79
perf,1,1024,1024,1,1,1,"wip",FWD_B,f32,0.0131836,159.073,0.0143948,145.688
perf,1,4096,1024,1,1,1,"wip",FWD_B,f32,0.0402832,208.241,0.0463605,180.943
perf,1,1024,4096,1,1,1,"wip",FWD_B,f32,0.0227051,369.46,0.0314383,266.827

# jit gemm, weights: oi
$ OMP_NUM_THREADS=28 KMP_HW_SUBSET=1s,28c,1t KMP_AFFINITY=compact,granularity=fine ./benchdnn --ip --mode=P --batch=inputs/ip/t.x
perf,42,3072,1024,1,1,1,"wip",FWD_B,f32,0.239014,1105.55,0.24625,1073.06
perf,42,1024,1024,1,1,1,"wip",FWD_B,f32,0.0991211,888.614,0.102301,860.991
perf,42,4096,1024,1,1,1,"wip",FWD_B,f32,0.347412,1014.13,0.35169,1001.8
perf,42,1024,4096,1,1,1,"wip",FWD_B,f32,0.377441,933.447,0.382852,920.255
perf,1,3072,1024,1,1,1,"wip",FWD_B,f32,0.137695,45.6911,0.143333,43.8941
perf,1,1024,1024,1,1,1,"wip",FWD_B,f32,0.0537109,39.0452,0.0565624,37.0768
perf,1,4096,1024,1,1,1,"wip",FWD_B,f32,0.202637,41.3973,0.210258,39.8967
perf,1,1024,4096,1,1,1,"wip",FWD_B,f32,0.25415,33.0065,0.270839,30.9727

# jit gemm, weights: io
$ OMP_NUM_THREADS=28 KMP_HW_SUBSET=1s,28c,1t KMP_AFFINITY=compact,granularity=fine ./benchdnn --ip --mode=P --batch=inputs/ip/t.x
perf,42,3072,1024,1,1,1,"wip",FWD_B,f32,0.175781,1503.24,0.182766,1445.79
perf,42,1024,1024,1,1,1,"wip",FWD_B,f32,0.0617676,1426,0.0682605,1290.36
perf,42,4096,1024,1,1,1,"wip",FWD_B,f32,0.634277,555.469,0.658485,535.049
perf,42,1024,4096,1,1,1,"wip",FWD_B,f32,0.219727,1603.45,0.232421,1515.87
perf,1,3072,1024,1,1,1,"wip",FWD_B,f32,0.0405273,155.24,0.0440751,142.744
perf,1,1024,1024,1,1,1,"wip",FWD_B,f32,0.0078125,268.435,0.0111665,187.808
perf,1,4096,1024,1,1,1,"wip",FWD_B,f32,0.0537109,156.181,0.06235,134.541
perf,1,1024,4096,1,1,1,"wip",FWD_B,f32,0.0620117,135.275,0.0669147,125.363
```

**Comment by vpirogov** [comment]
This change is targeting v1.0.

**Comment by vpirogov** [comment]
Default format for inner product changed to io in dff75ccab42fe4046f16746aad3cbd83e1e3fefc. The fix is available in v1.0-rc.
